src/util/xeval/builtin_cast.rs

```diff
@@ -0,0 +1,335 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use tipb::expression::Expr;
+use util::codec::datum::Datum;
+use util::codec::mysql::{Decimal, Time, Duration, DEFAULT_FSP};
+use super::{Evaluator, EvalContext, Result, Error};
+
+fn invalid_type_error(datum: &Datum) -> Result<Datum> {
+    Err(Error::Eval(format!("invalid expr type: {:?}, expect to be int", datum)))
+}
+
+fn check_datum_int(datum: &Datum) -> Result<()> {
+    match *datum {
+        Datum::I64(_) | Datum::U64(_) => Ok(()),
+        _ => Err(Error::Eval(format!("invalid expr type: {:?}, expect to be int", datum))),
+    }
+}
+
+impl Evaluator {
+    pub fn cast_int_as_int(&mut self, ctx: &EvalContext, expr: &Expr) -> Result<Datum> {
+        let child = try!(self.get_one_child(expr));
+        let d = try!(self.eval(ctx, child));
+        match d {
+            Datum::I64(_) => Ok(d),
+            Datum::U64(_) => Ok(Datum::I64(d.i64())),
```

Comment:
It seems U64::MAX would be converted to -1, while it would been converted to I64::MAX and return overflow error in tidb https://github.com/pingcap/tidb/blob/master/util/types/convert.go#L94